### ISSUE TITLE
accessing multiple fields in a recarray gives fields in wrong order (migrated from Trac #1431)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2984]
Original ticket http://projects.scipy.org/numpy/ticket/1431
Reported 2010-03-16 by atmention:samtygier, assigned to unknown.

when using the list in list notation to access multiple columns of a record array the columns are returned in the order that they are stored in the array, not in the order asked for.
    >>> x = np.array([(1.5,2.5,(1.0,2.0)),(3.,4.,(4.,5.)),(1.,3.,(2.,6.))],
            dtype=[('x','f4'),('y',np.float32),('value','f4',(2,2))])
    >>> x[['x','y']]
    array([(1.5, 2.5), (3.0, 4.0), (1.0, 3.0)],
         dtype=[('x', '<f4'), ('y', '<f4')])
    >>> x[['y','x']]
    array([(1.5, 2.5), (3.0, 4.0), (1.0, 3.0)],
         dtype=[('x', '<f4'), ('y', '<f4')])

this different to what happens with a 2d array, where they are returned in the order asked for:

```
>>> a = array([0,0.1,0.2,0.3,0.4])
then
>>> a[[0,1,4]]
array([ 0. ,  0.1,  0.4])
>>> a[[4,1,0]]
array([ 0.4,  0.1,  0. ])
```

I suggest that the it would be more consistent and predictable for the order to be the one asked for in both cases.

The current behaviour for the recarray was undocumented until today, so there can't be many people using it. therefore it might be worth break backwards compatibility to make the interface better.

Please see discussion at http://thread.gmane.org/gmane.comp.python.numeric.general/36933

a simple patch is attached.

**Comment by thouis** [comment]
Attachment in Trac by atmention:samtygier, 2010-03-16: [structured_array_index_order.diff](http://projects.scipy.org/numpy/attachment/ticket/1431/structured_array_index_order.diff)

**Comment by thouis** [comment]
Comment in Trac by atmention:samtygier, 2010-03-16

thinking about it the only case the would break is if some existing code asked for an order like cba, and expected to receive abc. i expect a programmer who asked for cba and found it did not work, would have changed their code to ask for abc, and handled the column order some other way.

**Comment by thouis** [comment]
Comment in Trac by atmention:samtygier, 2010-03-16

**Comment by thouis** [comment]
Comment in Trac by atmention:jseabold, 2010-06-18

Would the old behavior need to be deprecated?  If so, how do you deprecate in core/_internal?  I tried to add a 

```
from numpy.lib.utils import deprecate_with_doc
```

for a patch, but then importing numpy fails I believe because of a circular import.

**Comment by thouis** [comment]
Attachment in Trac by atmention:samtygier, 2010-06-19: [deprecate_old_order.diff](http://projects.scipy.org/numpy/attachment/ticket/1431/deprecate_old_order.diff)

**Comment by thouis** [comment]
Comment in Trac by atmention:samtygier, 2010-06-19

could just call warnings.warn() directly. we dont want to deprecate the whole function so i don't think deprecate_with_doc() is appropriate. here is a patch to warn you if you use you give fields out of order. it would only warn you if the behaviour would change by applying the first patch.

**Comment by thouis** [comment]
Attachment in Trac by atmention:samtygier, 2010-06-19: [rec_array_test.py](http://projects.scipy.org/numpy/attachment/ticket/1431/rec_array_test.py)

**Comment by thouis** [comment]
Comment in Trac by atmention:jseabold, 2010-06-19

Good point.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2010-08-05

Can you please check this?
http://github.com/rgommers/numpy/commit/0ad3bfd0ea86314466748f982b690f33a7b3a759

**Comment by thouis** [comment]
Comment in Trac by atmention:jseabold, 2010-08-05

Looks good to me.  Thanks for doing this!

```
In [2]: import numpy as np

In [3]: x = np.array([(1.5,2.5,(1.0,2.0)),(3.,4.,(4.,5.)),(1.,3.,(2.,6.))],
   ...:         dtype=[('x','f4'),('y',np.float32),('value','f4',(2,2))])

In [4]: 

In [5]: x[['y','x']]
/usr/local/lib/python2.6/dist-packages/numpy/core/_internal.py:344: DeprecationW
arning: Out of order field selection on recarrays currently returns fields in or
der. This behavior is deprecated in numpy 1.5 and will change in 2.0. See ticket
 #1431.
  warnings.warn(depdoc, DeprecationWarning)
```

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2010-08-14

Deprecated in r8626 and r8632. 

Checking deprecations is part of release process, and the test will complain if this is not fixed in 2.0. So we don't need this ticket anymore, therefore closing.
